.travis.yml

```diff
@@ -0,0 +1,70 @@
+language: cpp
+
+matrix:
+  include:
+    # Linux, default compilers
+    - os: linux
+    # Linux, latest supported compilers
+    - os: linux
+      env: GCC=gcc-6 CLANG=clang-6.0
+      addons:
+        apt:
+          sources:
+            - ubuntu-toolchain-r-test
+            - llvm-toolchain-trusty-6.0
+          packages:
+            - gcc-6
+            - g++-6
+            - clang-6.0
+    # Linux, AddressSanitizer
+    - os: linux
+      env: CFG=asan
+      sudo: true # LeakSanitizer doesn't work inside Travis containers, see https://github.com/travis-ci/travis-ci/issues/9033
+    # Linux, MemorySanitizer
+    - os: linux
+      env: CFG=msan
+    # OSX, default compilers, disabled because some tests in CoreTest are consistently failing and FBuildTest times out.
+    #- os: osx
+    #  compiler: clang
+  allow_failures:
+    - env: CFG=asan
+    - env: CFG=msan
+    - os: osx
+  fast_finish: true
+
+script:
+  - cd Code
+  # Set up some variables for convenience
+  - export NPROC=$(if [[ "${TRAVIS_OS_NAME}" == osx ]]; then sysctl -n hw.ncpu; else nproc; fi)
+  - export MAKE_CMD="make -j${NPROC}"
+  - export FBUILD_CMD="./fbuild -j${NPROC} -noprogress"
+  - export GCC=${GCC:-gcc} CLANG=${CLANG:-clang}
+  - export GXX=${GCC/gcc/g++} CLANGXX=${CLANG/clang/clang++}
+  # Patch configs to use installed compilers
+  - |
+    sed -ie "1i\\
+    #define CI_BUILD
+    s:CI_GCC_BINARY:$(which ${GCC}):
+    s:CI_GXX_BINARY:$(which ${GXX}):
+    s:CI_CC1_BINARY:$(${GCC} -print-prog-name=cc1):
+    s:CI_CC1PLUS_BINARY:$(${GCC} -print-prog-name=cc1plus):
+    s:CI_CLANG_BINARY:$(which ${CLANG}):
+    s:CI_CLANGXX_BINARY:$(which ${CLANGXX}):" fbuild.bff Tools/FBuild/FBuildTest/Data/testcommon.bff
+  # Bootstrap fbuild using Makefile
+  - ${MAKE_CMD} -s -f Makefile.bootstrap fbuild
```

Comment:
Is it possible to submit the binary for fbuild to the repo instead of bootstrapping like this, so that we can avoid the maintenance cost of maintaining the makefile?

Reply:
I think it would be the worst option, let me expand on this.

We have three ways of obtaining `fbuild` binary in the CI environment:
1. Download the binary from fastbuild.org.
    **Pros**:
    * Easy to implement.
    * Low (but continuous) maintenance: Version in the URL will need to be updated after each release, or when .bff files will actually start using features of a new release.

    **Cons**:
    * Builds will fail if CI will fail to download the file for some reason.
2. Store the binary to the repo.
    **Pros**:
    * Even easier to implement.
    * Low (but continuous) maintenance: Committed binary will need to be updated after each release, or when .bff files will actually start using features of a new release.

    **Cons**:
    * Doing so will significantly increase size of the repo over time. This is bad because everybody will now have to download and store these files that nobody besides CI need.
    For example: current size of the repo (`.git/objects` directory) is about 9Mb, size of the Linux binary is 4.2Mb, OSX binary - 1.4Mb, so after 2 releases the repo size will double.
3. Build the binary from source using other build system (make in this cake).
    **Pros**:
    * Already implemented :)
    * Because the makefile does a non-unity build we are testing non-unity builds for free.
    * Zero maintenance (most of the time): Proposed makefile can build any tagged version of FASTBuild starting from v0.80.

    **Cons**:
    * We are not testing that current sources can be built with a release version of FASTBuild. So if some new .bff feature (not included in any released FASTBuild version) will be used in the build configuration CI will not catch that.
    * Higher upper bound on maintenance costs: If structure of the source tree will change (more directory layers, new libraries, etc) or if some new mandatory compiler flag (like requiring newer C++ standard) will appear then the makefile will need to updated.
    This is definitely harder than incrementing a number in the URL or commiting a new file. But so far changes that would require an update happened only once (between 0.77 and 0.80).

Also having a way to build the `fbuild` binary that doesn't require a `fbuild` binary has some benefits not related to CI:
1. It enables initial build in environments where pre-built binary doen't fork (e.g. different libc implementation: uclibc or musl).
2. It enables initial build in situations when running binaries downloaded from the Internet is prohibited (e.g. company security policy).

Let me know what you think. If you still don't want the makefile, I think option 1 would be better.

Reply:
I think using pre-built binaries (either pulling them from fastbuild.org or checking them into the repo) would greatly reduce the complexity and maintenance of this change. Pulling them from fastbuild.org would only have to be updated 3-4 times a year at the current release rate.

For what it's worth, until now, the Linux and OSX executables have had no effort put into their size (so the Linux ones contained full debug info for example). As we approach v1.0, I've extracted the debug info from the Linux exe and will look into the size of the OSX exe soon. Your point about binaries bloating the gitrepor still stands of course, but this will be less of an issue with the Linux exe now comparable to the Windows one in size.

Reply:
> It does sound like option 1) is the way to go to me.

Changed build script to download the archive for the latest release from fastbuild.org